### PR TITLE
Findbugs high

### PR DESCRIPTION
**Pull request by ricellis**
*What*
Enabled findbugs reporting at high level

*How*
Applied findbugs plugin to gradle build for main source sets.
Fixed existing high issues:
* Avoided unnecessary unboxing of long
* Ensured the charset encoding is specified
* Fixed broken `equals` method and added `hashCode` methods
* Added excludes for class names shadowing superclasses (another option would be to rename these classes).
* Fixed error message formatting

*Testing*
Findbugs passes for local `gradlew clean build`.
Existing unit and integration tests pass, no new tests added.

reviewer @tomblench 
reviewer @rhyshort

### CODE REVIEW

**Comment by rhyshort** [comment]
+1

File: cloudant-sync-datastore-core/findbugs_excludes.xml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  -->
+<!--
+A list of known findbugs issues that we cannot fix at this time.
+-->
+<FindBugsFilter>
+
+    <!-- A couple of classes shadow the names of their supertypes.
+    -->
+    <Match>
+        <Bug code="Nm" pattern="NM_SAME_SIMPLE_NAME_AS_SUPERCLASS" />
```
Comment (tomblench) [inline cloudant-sync-datastore-core/findbugs_excludes.xml]: Should these be marked as a TODO or tracked in another ticket? I mean we're only excluding them to make findbugs pass, but we should fix the underlying problem eventually.

File: .travis.yml [hunk 2]
```diff
@@ -3,7 +3,7 @@ sudo: false
 script: "./gradlew check integrationTest"
 
 jdk:
-  - openjdk6
+  - openjdk7
```
Comment (tomblench) [inline .travis.yml]: So openjdk 7 is required for findbugs? I guess this is a reasonable upgrade, do we think there are any risks of compiling/running under a newer jdk?

Comment (rhyshort) [inline .travis.yml]: The only risk is using APIs and language features not in 6, the android builds on jenkins should pick that up though

Comment (ricellis) [inline .travis.yml]: Yes, findbugs requires 1.7 to run. I guess the risk is that we accidentally include some Java 7 API, but using the correct Java level in the IDE project settings and the Android testing should help protect us from that.

File: cloudant-sync-datastore-core/findbugs_excludes.xml (hunk 1 above)
Comment (ricellis) [inline cloudant-sync-datastore-core/findbugs_excludes.xml]: Raised #252 and added a reference to it in the excludes in 45cc5f2.

**Comment by tomblench** [comment]
+1

**Pull request merged**